kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ReverseFilesToColumnarBatchActionsIterable.java

```diff
@@ -0,0 +1,247 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.internal.replay;
+
+import io.delta.kernel.client.FileReadContext;
+import io.delta.kernel.client.JsonHandler;
+import io.delta.kernel.client.ParquetHandler;
+import io.delta.kernel.client.TableClient;
+import io.delta.kernel.data.FileDataReadResult;
+import io.delta.kernel.expressions.Literal;
+import io.delta.kernel.fs.FileStatus;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.lang.CloseableIterable;
+import io.delta.kernel.internal.util.InternalUtils;
+import io.delta.kernel.internal.util.Logging;
+import io.delta.kernel.types.StructType;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.Tuple2;
+import io.delta.kernel.utils.Utils;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class takes as input a list of delta files (.json, .checkpoint.parquet) and produces an
+ * iterator of (FileDataReadResult, isFromCheckpoint) tuples.
+ *
+ * Users can pass in a `readSchema` to select which actions they want to consume.
+ */
+public class ReverseFilesToColumnarBatchActionsIterable
+    implements CloseableIterable<Tuple2<FileDataReadResult, Boolean>>, Logging {
+
+    private final TableClient tableClient;
+    private final List<FileStatus> reverseSortedFiles;
+    private final StructType readSchema;
+
+    public ReverseFilesToColumnarBatchActionsIterable(
+            TableClient tableClient,
+            List<FileStatus> filesUnsorted,
+            StructType readSchema) {
+        this.tableClient = tableClient;
+        this.reverseSortedFiles = filesUnsorted
+            .stream()
+            .sorted(
+                Comparator
+                    .comparing((FileStatus a) -> new Path(a.getPath()).getName())
+                    .reversed()
+            ).collect(Collectors.toList());
+        this.readSchema = readSchema;
+    }
+
+    @Override
+    public CloseableIterator<Tuple2<FileDataReadResult, Boolean>> iterator() {
+        return new CloseableIterator<Tuple2<FileDataReadResult, Boolean>>() {
+            /**
+             * Iterator over the files, in reverse.
+             *
+             * Each file will be split (by 1, or more) to yield an iterator of FileDataReadResults.
+             */
+            private final Iterator<FileStatus> filesIter = reverseSortedFiles.iterator();
+
+            /**
+             * The current (FileDataReadResult, isFromCheckpoint) tuple. Whenever this iterator
+             * is exhausted, we will try and fetch the next one from the `filesIter`.
+             */
+            private Optional<CloseableIterator<Tuple2<FileDataReadResult, Boolean>>>
+                actionsIter = Optional.empty();
+
+            private boolean closed = false;
```

Comment:
nit: default initialization value is `false`, no need to initialize

Reply:
Actually, for local variables, you do need to initialize them btw:

https://web.archive.org/web/20140814175549/https://docs.oracle.com/javase/specs/jls/se5.0/html/typesValues.html#96595

A local variable ([§14.4](https://web.archive.org/web/20140814175549/https://docs.oracle.com/javase/specs/jls/se5.0/html/statements.html#5920), [§14.14](https://web.archive.org/web/20140814175549/https://docs.oracle.com/javase/specs/jls/se5.0/html/statements.html#24588)) must be explicitly given a value before it is used, by either initialization [(§14.4)](https://web.archive.org/web/20140814175549/https://docs.oracle.com/javase/specs/jls/se5.0/html/statements.html#5920) or assignment [(§15.26)](https://web.archive.org/web/20140814175549/https://docs.oracle.com/javase/specs/jls/se5.0/html/expressions.html#5281), in a way that can be verified by the compiler using the rules for definite assignment [(§16)](https://web.archive.org/web/20140814175549/https://docs.oracle.com/javase/specs/jls/se5.0/html/defAssign.html#25979).